sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -470,3 +397,451 @@ object JoinReorderDPFilters extends PredicateHelper {
+        newGenes :+= current
+      } else if (filteredTable.nonEmpty) {
+        current = filteredTable.toIndexedSeq(util.Random.nextInt(filteredTable.size))._2.head
+        newGenes :+= current
+      } else {
+        stop = true
+      }
+    }
+
+    Chromosome(father.conf, newGenes, father.conditions, father.topOutputSet)
+  }
+}
+
+/**
+ * A sequence of genes(each represents a single relation) which represents
+ * a joined plan with determined join order.
+ */
+case class Chromosome(
+    conf: SQLConf,
+    basicPlans: Seq[JoinPlan],
+    conditions: Set[Expression],
+    topOutputSet: AttributeSet) {
+
+  lazy val fitness: Double = evalFitness(integratedPlan)
+
+  lazy val integratedPlan: Option[JoinPlan] = makePlan
+
+  private def makePlan: Option[JoinPlan] = {
+    val semiFinished = mutable.Buffer[JoinPlan]()
+    basicPlans.foreach(mergeSemi(semiFinished, _))
+    if (semiFinished.head.itemIds.size == basicPlans.size) {
+      Some(semiFinished.head)
+    } else {
+      None
+    }
+  }
+
+  private def mergeSemi(semiFinished: mutable.Buffer[JoinPlan], right: JoinPlan): Unit = {
+    val filters = None: Option[JoinGraphInfo]
+    for (left <- semiFinished) {
+      JoinReorderUtils.buildJoin(left, right, conf, conditions, topOutputSet, filters) match {
+        case Some(joined) =>
+          semiFinished.remove(semiFinished.indexOf(left))
+          mergeSemi(semiFinished, joined)
+        case _ =>
+          None
+      }
+    }
+
+    if (semiFinished.isEmpty || right.itemIds.size == 1) {
+      semiFinished.append(right)
+      return
+    }
+
+    insertPlan(semiFinished, right)
+  }
+
+  private def insertPlan(semiFinished: mutable.Buffer[JoinPlan], plan: JoinPlan): Unit = {
+    var criticalSize = if (semiFinished.head.itemIds.size > plan.itemIds.size) {
+      semiFinished.head.itemIds.size
+    } else {
+      plan.itemIds.size
+    }
+    var criticalIndex = 0
+    var break: Boolean = false
+    for (p <- semiFinished if !break) {
+      if (plan.itemIds.size > p.itemIds.size && plan.itemIds.size <= criticalSize) {
+        break = true
+      } else {
+        criticalIndex += 1
+        criticalSize = p.itemIds.size
+      }
+    }
+
+    semiFinished.insert(criticalIndex, plan)
+  }
+
+  private def evalFitness(plan: Option[JoinPlan]): Double = {
+    plan match {
+      case Some(joinPlan) =>
+        // We use the negative cost as fitness.
+        - joinPlan.planCost.card.toDouble * conf.joinReorderCardWeight -
+            joinPlan.planCost.size.toDouble * (1 - conf.joinReorderCardWeight)
+      case _ =>
+        - Double.MaxValue
+    }
+  }
+}
+
+/**
+ * A live space which has multi individuals(represented by [[Chromosome]]).
+ * The population can evolve, generation by generation. The child individual
+ * is generated by the [[Crossover]] procedure from its parents.
+ */
+object Population {
+  def apply(
+      conf: SQLConf,
+      itemsMap: Map[Int, JoinPlan],
```

Comment:
I wonder do we really need an `itemsMap` here? It's necessary in `JoinReorderDP` cuz it needs to maintain subset relationships. However, in `JoinReorderGA` you have a population of orderings in the first place.

Reply:
Yes, you're right. What we need here is just the plans. Have updated the code.